File: sidecar/src/services/cron.ts
import {
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  StringSelectMenuBuilder,
  EmbedBuilder,
} from "discord.js";
import type { DiscordBridgeClient } from "./bridge.js";
import type { CronJobItem, InboundCronEvent } from "../types.js";

function truncate(str: string, maxLen: number): string {
  if (!str) return "";
  return str.length > maxLen ? `${str.slice(0, maxLen - 3)}...` : str;
}

export class CronManager {
  private jobsCache = new Map<string, CronJobItem>();

  constructor(private bridge: DiscordBridgeClient) {}

  public setJobs(jobs: CronJobItem[]): void {
    this.jobsCache.clear();
    for (const job of jobs) {
      this.jobsCache.set(job.id, job);
    }
  }

  public getJobs(): CronJobItem[] {
    return Array.from(this.jobsCache.values());
  }

  public getJob(id: string): CronJobItem | undefined {
    return this.jobsCache.get(id);
  }

  public updateJob(job: CronJobItem): void {
    this.jobsCache.set(job.id, job);
  }

  public removeJob(id: string): void {
    this.jobsCache.delete(id);
  }

  public requestList(chatId: string, senderId: string, interactionId?: string): void {
    const event: InboundCronEvent = {
      type: "cron",
      platform: "discord",
      action: "list",
      chat_id: chatId,
      sender_id: senderId,
      interaction_id: interactionId,
    };
    this.bridge.emitInboundEvent(event);
  }

  public requestAction(
    action: "pause" | "resume" | "trigger" | "delete" | "view_output",
    jobId: string,
    chatId: string,
    senderId: string,
    interactionId?: string
  ): void {
    const event: InboundCronEvent = {
      type: "cron",
      platform: "discord",
      action,
      job_id: jobId,
      chat_id: chatId,
      sender_id: senderId,
      interaction_id: interactionId,
    };
    this.bridge.emitInboundEvent(event);
  }

  public requestCreate(
    name: string,
    schedule: string,
    prompt: string,
    deliver: string,
    chatId: string,
    senderId: string,
    interactionId?: string
  ): void {
    const event: InboundCronEvent = {
      type: "cron",
      platform: "discord",
      action: "create",
      name,
      schedule,
      prompt,
      deliver,
      chat_id: chatId,
      sender_id: senderId,
      interaction_id: interactionId,
    };
    this.bridge.emitInboundEvent(event);
  }

  /**
   * Construction de l'Embed de liste des tâches cron.
   */
  public buildJobListEmbed(jobs: CronJobItem[]): EmbedBuilder {
    const embed = new EmbedBuilder()
      .setTitle("⏰ Gestionnaire des Tâches Cron Hermes")
      .setColor(0x5865f2)
      .setDescription(
        `Consultez et pilotez vos tâches planifiées ci-dessous.\nUtilisez le menu pour inspecter une tâche, ou cliquez sur **➕ Planifier une tâche** pour en ajouter une nouvelle.\n\n` +
          (jobs.length === 0
            ? "*Aucune tâche cron configurée pour le moment.*"
            : jobs
                .slice(0, 10)
                .map((j, idx) => {
                  const icon = j.enabled ? "🟢" : "⏸️";
                  const sched = j.schedule_display || j.schedule;
                  const nextStr = j.next_run
                    ? typeof j.next_run === "number"
                      ? `<t:${Math.floor(j.next_run / 1000)}:R>`
                      : j.next_run
                    : "Non planifié";
                  return `**${idx + 1}.** ${icon} **${truncate(j.name, 35)}** (\`${sched}\`)\n↳ Prochaine : ${nextStr}`;
                })
                .join("\n\n"))
      )
      .setFooter({ text: `${jobs.length} tâche(s) enregistrée(s)` })
      .setTimestamp();

    return embed;
  }

  /**
   * Construction des composants pour la liste des tâches cron.
   */
  public buildJobListComponents(jobs: CronJobItem[]): ActionRowBuilder<StringSelectMenuBuilder | ButtonBuilder>[] {
    const rows: ActionRowBuilder<StringSelectMenuBuilder | ButtonBuilder>[] = [];

    if (jobs.length > 0) {
      const selectMenu = new StringSelectMenuBuilder()
        .setCustomId("cron:select")
        .setPlaceholder("🔍 Choisissez une tâche cron à piloter...")
        .setMinValues(1)
        .setMaxValues(1)
        .addOptions(
          jobs.slice(0, 25).map((j) => ({
            label: truncate(j.name, 90),
            value: j.id,
            description: truncate(
              `${j.enabled ? "🟢 Active" : "⏸️ En pause"} • Sched: ${j.schedule_display || j.schedule}`,
              100
            ),
            emoji: j.enabled ? "🟢" : "⏸️",
          }))
        );
      rows.push(new ActionRowBuilder<StringSelectMenuBuilder>().addComponents(selectMenu));
    }

    const controlRow = new ActionRowBuilder<ButtonBuilder>().addComponents(
      new ButtonBuilder()
        .setCustomId("cron:create_modal")
        .setLabel("Planifier une tâche")
        .setStyle(ButtonStyle.Success)
        .setEmoji("➕"),
      new ButtonBuilder()
        .setCustomId("cron:refresh")
        .setLabel("Rafraîchir")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji("🔄"),
      new ButtonBuilder()
        .setCustomId("cron:dismiss")
        .setLabel("Fermer")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji("✖️")
    );
    rows.push(controlRow);

    return rows;
  }

  /**
   * Construction de l'Embed détaillé d'une tâche cron.
   */
  public buildJobDetailEmbed(job: CronJobItem): EmbedBuilder {
    const statusText = job.enabled ? "🟢 Active" : "⏸️ En pause";
    const schedText = job.schedule_display || job.schedule;

    const nextStr = job.next_run
      ? typeof job.next_run === "number"
        ? `<t:${Math.floor(job.next_run / 1000)}:F> (<t:${Math.floor(job.next_run / 1000)}:R>)`
        : job.next_run
      : "Non calculé";

    const lastStr = job.last_run
      ? typeof job.last_run === "number"
        ? `<t:${Math.floor(job.last_run / 1000)}:R>`
        : job.last_run
      : "Jamais exécuté";

    const lastStatus = job.last_status === "success"
      ? "✅ Succès"
      : job.last_status === "error"
      ? "❌ Erreur"
      : job.last_status || "N/A";

    const embed = new EmbedBuilder()
      .setTitle(`⏰ Tâche Cron : ${truncate(job.name, 60)}`)
      .setColor(job.enabled ? 0x57f287 : 0xfee75c)
      .setDescription(`**Prompt / Instruction exécutée :**\n\`\`\`text\n${truncate(job.prompt, 1500)}\n\`\`\``)
      .addFields(
        { name: "📋 Statut", value: `**${statusText}**`, inline: true },
        { name: "🗓️ Fréquence / Cron", value: `\`${schedText}\``, inline: true },
        { name: "📍 Destination", value: `\`${job.deliver || "Automatique"}\``, inline: true },
        { name: "⏱️ Prochaine exécution", value: String(nextStr), inline: false },
        { name: "📜 Dernier résultat", value: `${lastStatus} (${lastStr})`, inline: false },
        { name: "🆔 Identifiant unique", value: `\`${job.id}\``, inline: false }
      )
      .setFooter({ text: "Actions disponibles ci-dessous" })
      .setTimestamp();

    return embed;
  }

  /**
   * Construction des boutons d'actions pour une tâche sélectionnée.
   */
  public buildJobDetailComponents(job: CronJobItem): ActionRowBuilder<ButtonBuilder>[] {
    const row1 = new ActionRowBuilder<ButtonBuilder>().addComponents(
      new ButtonBuilder()
        .setCustomId(`cron:trigger:${job.id}`)
        .setLabel("Exécuter maintenant")
        .setStyle(ButtonStyle.Primary)
        .setEmoji("⚡"),
      new ButtonBuilder()
        .setCustomId(`cron:toggle_pause:${job.id}`)
        .setLabel(job.enabled ? "Mettre en pause" : "Reprendre")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji(job.enabled ? "⏸️" : "▶️"),
      new ButtonBuilder()
        .setCustomId(`cron:view_output:${job.id}`)
        .setLabel("Dernier output")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji("📜")
    );

    const row2 = new ActionRowBuilder<ButtonBuilder>().addComponents(
      new ButtonBuilder()
        .setCustomId(`cron:delete_prompt:${job.id}`)
        .setLabel("Supprimer la tâche")
        .setStyle(ButtonStyle.Danger)
        .setEmoji("🗑️"),
      new ButtonBuilder()
        .setCustomId("cron:list")
        .setLabel("Retour à la liste")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji("⬅️")
    );

    return [row1, row2];
  }

  /**
   * Construction des boutons de confirmation de suppression d'une tâche.
   */
  public buildDeleteConfirmationComponents(jobId: string): ActionRowBuilder<ButtonBuilder>[] {
    const row = new ActionRowBuilder<ButtonBuilder>().addComponents(
      new ButtonBuilder()
        .setCustomId(`cron:delete_confirm:${jobId}`)
        .setLabel("Confirmer la suppression")
        .setStyle(ButtonStyle.Danger)
        .setEmoji("💥"),
      new ButtonBuilder()
        .setCustomId(`cron:detail:${jobId}`)
        .setLabel("Annuler")
        .setStyle(ButtonStyle.Secondary)
        .setEmoji("↩️")
    );
    return [row];
  }
}

File: views/cron.py
"""Gestionnaire et vues interactives pour les tâches Cron (/cron)."""

from __future__ import annotations

import logging
from typing import Any, Dict, List, Optional

import discord
from discord import Embed, TextStyle
from discord.ui import Button, Modal, Select, TextInput, View

from .common import truncate

logger = logging.getLogger(__name__)

# =============================================================================

class CronCreateModal(Modal, title="Planifier une nouvelle tâche Cron"):
    name = TextInput(label="Nom de la tâche", placeholder="ex: Veille Matinale, Rapport Git...", required=True)
    schedule = TextInput(label="Fréquence / Expression Cron", placeholder="ex: 0 9 * * 1-5, every 2h...", required=True)
    prompt = TextInput(label="Instructions / Prompt", style=TextStyle.paragraph, placeholder="Consigne détaillée exécutée par l'agent...", required=True)
    deliver = TextInput(label="Salon de livraison (optionnel)", placeholder="Identifiant de salon ou 'dm' (vide = salon actuel)", required=False)

    def __init__(self, adapter: Any, chat_id: str):
        super().__init__()
        self.adapter = adapter
        self.chat_id = chat_id

    async def on_submit(self, interaction: discord.Interaction):
        await interaction.response.send_message(
            f"⏳ Création de la tâche cron **{self.name.value}** en cours...", ephemeral=True
        )
        await self.adapter.handle_cron_create(
            name=self.name.value.strip(),
            schedule=self.schedule.value.strip(),
            prompt=self.prompt.value.strip(),
            deliver=self.deliver.value.strip() or self.chat_id,
            chat_id=self.chat_id,
        )


class CronSelect(Select):
    def __init__(self, view: CronManagerView, jobs: List[Dict[str, Any]]):
        self.cron_view = view
        opts = []
        for j in jobs[:25]:
            icon = "🟢" if j.get("enabled", True) else "⏸️"
            lbl = truncate(j.get("name") or j.get("id", "Job"), 100)
            sched = truncate(j.get("schedule_display") or j.get("schedule", ""), 100)
            opts.append(
                discord.SelectOption(
                    label=f"{icon} {lbl}",
                    value=str(j.get("id")),
                    description=sched,
                )
            )
        super().__init__(placeholder="Inspecter une tâche cron...", min_values=1, max_values=1, options=opts)

    async def callback(self, interaction: discord.Interaction):
        job_id = self.values[0]
        await self.cron_view.show_job_detail(interaction, job_id)


class CronManagerView(View):
    """Vue interactive pour la commande /cron."""

    def __init__(self, adapter: Any, chat_id: str, jobs: List[Dict[str, Any]]):
        super().__init__(timeout=300.0)
        self.adapter = adapter
        self.chat_id = chat_id
        self.jobs = jobs
        self.selected_job_id: Optional[str] = None

        if jobs:
            self.add_item(CronSelect(self, jobs))

    def build_list_embed(self) -> Embed:
        embed = Embed(
            title="⏰ Gestionnaire des Tâches Cron Hermes",
            color=0x5865F2,
            description=(
                "Consultez et pilotez vos tâches planifiées ci-dessous.\n"
                "Sélectionnez une tâche dans le menu pour l'inspecter, ou cliquez sur **➕ Planifier**.\n\n"
            ),
        )
        if not self.jobs:
            embed.description += "*Aucune tâche cron configurée pour le moment.*"
        else:
            for idx, j in enumerate(self.jobs[:10]):
                icon = "🟢" if j.get("enabled", True) else "⏸️"
                sched = j.get("schedule_display") or j.get("schedule", "")
                embed.add_field(
                    name=f"{idx + 1}. {icon} {truncate(j.get('name') or j.get('id'), 40)}",
                    value=f"↳ Fréquence : `{sched}`",
                    inline=False,
                )
        return embed

    def build_detail_embed(self, job: Dict[str, Any]) -> Embed:
        is_enabled = job.get("enabled", True)
        embed = Embed(
            title=f"{'🟢' if is_enabled else '⏸️'} Tâche : {job.get('name') or job.get('id')}",
            color=0x57F287 if is_enabled else 0xFEE75C,
        )
        embed.add_field(name="Identifiant", value=f"`{job.get('id')}`", inline=True)
        embed.add_field(name="Statut", value="Actif" if is_enabled else "En pause", inline=True)
        embed.add_field(name="Fréquence", value=f"`{job.get('schedule_display') or job.get('schedule')}`", inline=True)
        if job.get("prompt"):
            embed.add_field(name="Prompt", value=f"```\n{truncate(job['prompt'], 800)}\n```", inline=False)
        return embed

    async def show_job_detail(self, interaction: discord.Interaction, job_id: str):
        self.selected_job_id = job_id
        job = next((j for j in self.jobs if str(j.get("id")) == job_id), None)
        if not job:
            await interaction.response.send_message("❌ Tâche introuvable.", ephemeral=True)
            return

        detail_view = CronDetailView(self.adapter, self.chat_id, job, self)
        await interaction.response.edit_message(embed=self.build_detail_embed(job), view=detail_view)

    @discord.ui.button(label="Planifier une tâche", style=discord.ButtonStyle.success, emoji="➕")
    async def create_button(self, interaction: discord.Interaction, button: Button):
        if not self.adapter.is_user_authorized(interaction.user.id):
            await interaction.response.send_message("⛔ Non autorisé.", ephemeral=True)
            return
        await interaction.response.send_modal(CronCreateModal(self.adapter, self.chat_id))

    @discord.ui.button(label="Rafraîchir", style=discord.ButtonStyle.secondary, emoji="🔄")
    async def refresh_button(self, interaction: discord.Interaction, button: Button):
        self.jobs = self.adapter.fetch_cron_jobs()
        self.clear_items()
        if self.jobs:
            self.add_item(CronSelect(self, self.jobs))
        self.add_item(self.create_button)
        self.add_item(self.refresh_button)
        await interaction.response.edit_message(embed=self.build_list_embed(), view=self)


class CronDetailView(View):
    """Boutons de contrôle pour une tâche cron spécifique."""

    def __init__(self, adapter: Any, chat_id: str, job: Dict[str, Any], parent_view: CronManagerView):
        super().__init__(timeout=300.0)
        self.adapter = adapter
        self.chat_id = chat_id
        self.job = job
        self.parent_view = parent_view
        self.job_id = str(job.get("id"))

        is_enabled = job.get("enabled", True)
        if is_enabled:
            self.pause_btn = Button(label="Mettre en pause", style=discord.ButtonStyle.secondary, emoji="⏸️")
            self.pause_btn.callback = self._pause_callback
            self.add_item(self.pause_btn)
        else:
            self.resume_btn = Button(label="Reprendre", style=discord.ButtonStyle.success, emoji="▶️")
            self.resume_btn.callback = self._resume_callback
            self.add_item(self.resume_btn)

    async def _pause_callback(self, interaction: discord.Interaction):
        await self.adapter.handle_cron_action("pause", self.job_id, self.chat_id)
        await interaction.response.send_message(f"⏸️ Tâche `{self.job_id}` mise en pause.", ephemeral=True)

    async def _resume_callback(self, interaction: discord.Interaction):
        await self.adapter.handle_cron_action("resume", self.job_id, self.chat_id)
        await interaction.response.send_message(f"▶️ Tâche `{self.job_id}` reprise.", ephemeral=True)

    @discord.ui.button(label="Exécuter maintenant", style=discord.ButtonStyle.primary, emoji="⚡")
    async def run_button(self, interaction: discord.Interaction, button: Button):
        await self.adapter.handle_cron_action("trigger", self.job_id, self.chat_id)
        await interaction.response.send_message(f"⚡ Exécution immédiate lancée pour `{self.job_id}`.", ephemeral=True)

    @discord.ui.button(label="Dernier output", style=discord.ButtonStyle.secondary, emoji="📋")
    async def output_button(self, interaction: discord.Interaction, button: Button):
        output = self.adapter.get_cron_last_output(self.job_id)
        if output:
            msg = f"📋 **Dernier rapport d'exécution pour `{self.job_id}` :**\n```markdown\n{output[:1800]}\n```"
        else:
            msg = f"ℹ️ Aucun rapport d'exécution disponible pour `{self.job_id}`."
        await interaction.response.send_message(msg, ephemeral=True)

    @discord.ui.button(label="Supprimer", style=discord.ButtonStyle.danger, emoji="🗑️")
    async def delete_button(self, interaction: discord.Interaction, button: Button):
        await self.adapter.handle_cron_action("delete", self.job_id, self.chat_id)
        await interaction.response.send_message(f"🗑️ Tâche `{self.job_id}` supprimée.", ephemeral=True)

    @discord.ui.button(label="Retour à la liste", style=discord.ButtonStyle.secondary, emoji="⬅️")
    async def back_button(self, interaction: discord.Interaction, button: Button):
        self.parent_view.jobs = self.adapter.fetch_cron_jobs()
        await interaction.response.edit_message(embed=self.parent_view.build_list_embed(), view=self.parent_view)
